### PR TITLE
feat: alternative way to bootstrap the tests

### PR DESCRIPTION
**Pull request by thall**
### TL;DR

With this setup you need to explicit **opt-out** tests instead of explicit **opt-int**.

### Why

We have noticed historically, that when people adds new
resources, it's easy to forget to implement the newly
generated AIP tests, either directly or later.

### How

This commit tries to solve that issue by introducing a
main entrypoint to execute the tests, which takes a interface
that the user needs to implement.

When a new resource is added, the interface will be extended with
another method that the user needs to implement, if not, a compilation
error would be raised.

The user can choose to return `nil` to indicate that it can't be
implemented right directly.

All tests will still be executed, but if not implemented it will be
skipped, this to show it's available but not implemented.

This commit is backwards compatible with the current test setup.

For example usage, see `examples/`.

### CODE REVIEW

**Review by fredrikaverpil — APPROVED** [review]
I'm not sure where this is headed, but it LGTM technically.

Ping us if you want to show where you are going with this 😄 

EDIT: never mind! I'm tired... somehow missed the last commit, which contains all the goodies.

**Review by fredrikaverpil — DISMISSED** [review]
I thought it was enough to hit re-review to retract the approval ... but no.

**Review by radhus — APPROVED** [review]
Nice to have an alternative interface! This is non-breaking (except users will get changes in their generated code, expected), so I think we should try it out!

File: proto/gen/googleapis/aiplatform/apiv1/aiplatformpb/dataset_service_aiptest.pb.go [hunk 1]
```diff
@@ -15,6 +15,123 @@ import (
...
+// DatasetServiceTestSuiteConfigProvider is the interface to implement to decide which resources
+// that should be tested and configured.
+type DatasetServiceTestSuiteConfigProvider interface {
+	AnnotationTestSuiteConfig(t *testing.T) *DatasetServiceAnnotationTestSuiteConfig
+	AnnotationSpecTestSuiteConfig(t *testing.T) *DatasetServiceAnnotationSpecTestSuiteConfig
+	DataItemTestSuiteConfig(t *testing.T) *DatasetServiceDataItemTestSuiteConfig
+	DatasetTestSuiteConfig(t *testing.T) *DatasetServiceDatasetTestSuiteConfig
+	DatasetVersionTestSuiteConfig(t *testing.T) *DatasetServiceDatasetVersionTestSuiteConfig
+	SavedQueryTestSuiteConfig(t *testing.T) *DatasetServiceSavedQueryTestSuiteConfig
+}
```
Comment (ericwenn) [inline proto/gen/googleapis/aiplatform/apiv1/aiplatformpb/dataset_service_aiptest.pb.go:27]: IMO add comment that returning nil from these methods is valid, and that it means the test is skipped.

File: example/freight_service_test.go [hunk 2]
```diff
@@ -51,3 +51,26 @@ func Test_FreightService(t *testing.T) {
...
+}
+
+func (a aipTests) SiteTestSuiteConfig(_ *testing.T) *examplefreightv1.FreightServiceSiteTestSuiteConfig {
+	// Since the service isn't implemented, no proper configuration can be given.
+	// The configuration is used as shown above, the only addition is the Service and Context methods.
+	return &examplefreightv1.FreightServiceSiteTestSuiteConfig{
+		Service: nil, // No service can be provided since it's not implemented.
+		Context: context.Background,
+	}
+}
```
Comment (ericwenn) [inline example/freight_service_test.go:76]: I don't understand this example, can't it be roughly the same as the original one above?

**Review by ericwenn — APPROVED** [review]
I like this!
Address the comment and example comments, and I think this is safe enough to try. I really like that it is not a breaking change for users (except that more code is generated) :star:

File: example/freight_service_test.go (hunk 2 above)
Comment (thall) [reply example/freight_service_test.go:76]: The outcome is the same, that the tests will be skipped, but for different reasons, this one will fail because no service was provided.
But I can simplify it and have the same as above.

**Comment by fredrikaverpil** [comment]
@niclas late to the party, but I was just wondering if you wanted to also add a note to the main README about this new type of setup?

**Comment by thall** [comment]
@fredrikaverpil yes, I will update the README as well, thanks for the reminder!
